martha/martha_v3.js

```diff
@@ -6,156 +6,207 @@ const url = require('url');
 const BAD_REQUEST_ERROR_CODE = 400;
 const SERVER_ERROR_CODE = 500;
 
+const BOND_PROVIDER_NONE = null; // Used for servers that should NOT contact bond
+const BOND_PROVIDER_DCF_FENCE = 'dcf-fence'; // The default when we don't recognize the server
+const BOND_PROVIDER_FENCE = 'fence';
+const BOND_PROVIDER_ANVIL = 'anvil';
+
+const AUTH_REQUIRED = true;
+const AUTH_SKIPPED = false;
+
+const PROTOCOL_PREFIX_DOS='/ga4gh/dos/v1/dataobjects';
+const PROTOCOL_PREFIX_DRS='/ga4gh/drs/v1/objects';
+
+// TODO: Does dataguids.org actually have a resolution API??
+// Until we figure that out, we'll try to re-implement the mappings we know about here.
+const DG_EXPANSION_NONE = null;
+const DG_EXPANSION_BIO_DATA_CATALYST = config.dataObjectResolutionHost;
+const DG_EXPANSION_THE_ANVIL = 'gen3.theanvil.io';
+
 class DrsType {
-    constructor(drsUrl, sendAuth, bondUrl, responseParser) {
-        this.drsUrl = drsUrl;
+    constructor(dataGuidExpansion, protocolPrefix, sendAuth, bondProvider) {
+        this.dataGuidExpansion = dataGuidExpansion;
+        this.protocolPrefix = protocolPrefix;
         this.sendAuth = sendAuth;
-        this.bondUrl = bondUrl;
-        this.responseParser = responseParser;
+        this.bondProvider = bondProvider;
     }
 }
 
 /** *************************************************************************************************
  * URI parsers
  */
 
-function dosFullUrlGenerator (parsedUrl) {
+function fullUrlGenerator (parsedUrl, protocolPrefix) {
     return url.format({
         protocol: 'https',
         hostname: parsedUrl.hostname,
         port: parsedUrl.port,
-        pathname: `/ga4gh/dos/v1/dataobjects${(parsedUrl.pathname) ? parsedUrl.pathname : `/${parsedUrl.hostname}`}`,
+        pathname: `${protocolPrefix}${parsedUrl.pathname || `/${parsedUrl.hostname}`}`,
         search: parsedUrl.search
     });
 }
 
-function dosCompactUrlGenerator (parsedUrl) {
-    const splitHost = parsedUrl.hostname.split('.');
-    let idPrefix;
-    if (parsedUrl.pathname) {
-        idPrefix = `${splitHost[0]}.${splitHost[1].toUpperCase()}`;
-    } else {
-        idPrefix = `${splitHost[0]}.${splitHost[1]}`;
-    }
-
+const compactUrlGenerator = (dataGuidExpansion) => function (parsedUrl, protocolPrefix) {
```

Comment:
why is this function declaration in this format instead of
```suggestion
function compactUrlGenerator (parsedUrl, dataGuidExpansion, protocolPrefix) {
```
like `httpsUrlGenerator` ?

Reply:
Because of how it's called. I (semi-)[curried the function](https://medium.com/javascript-scene/curry-and-function-composition-2c208d774983) and then only partially applied it before as its used below.